### PR TITLE
fix (#minor); makerdao-ethereum; fix crash due to negative position bal reported in #1855

### PR DESCRIPTION
**Pull request by tnkrxyz**
This PR fixes crash of makerdao subgraph (#1855) due to negative borrower balance in the D3MHub because the subgraph didn't account for fees in D3MHub contract.

### deployment & dashboard (drafting/sync in progress)
- deployment 1: https://thegraph.com/hosted-service/subgraph/tnkrxyz/makerdao-ethereum?version=pending (drafted from messari/makerdao-ethereum)
- deployment 2: https://thegraph.com/hosted-service/subgraph/tnkrxyz/test-subgraph (sync from scratch)
- dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/id/QmeVJZYjMnmL3KUQXAJjkFLLXFnnwhnz8SNxSj96fQ2A5y&tab=protocol

### CODE REVIEW

File: subgraphs/makerdao/src/common/helpers.ts [hunk 1]
```diff
@@ -510,10 +509,18 @@ export function updatePosition(
         PositionSide.LENDER,
         true
       );
-      assert(
-        deltaCollateral.gt(BIGINT_ZERO),
-        `[updatePosition]Creating new lender position ${lenderPosition.id} with deltaCollateral ${deltaCollateral} <= 0`
-      );
+
+      if (deltaCollateral.le(BIGINT_ZERO)) {
+        log.critical(
```
Comment (jaimehgb) [inline subgraphs/makerdao/src/common/helpers.ts:514]: From what I've seen, `log.criticals` don't output the message to the logs. Instead do a `log.error` with all the data you want to log, followed by a `log.critical` to then stop the execution

File: subgraphs/makerdao/src/common/helpers.ts [hunk 2]
```diff
@@ -611,31 +626,23 @@ export function updatePosition(
...
-    ) {
-      log.warning(
-        "[updatePosition]A small negative borrow balance of {} for position {} with tx {}; deemed position closed (balance set to 0)",
+    // this may be less than 0, but not too negative (>-100), e.g.
+    // 1. at block 12507581, urn 0x03453d22095c0edd61cd40c3ccdc394a0e85dc1a
+    // repaid -203334964101176257798573 dai, when the borrow balance
+    // was 203334964101176257798572
+    // 2. at block 14055178, urn 0x1c47bb6773db2a441264c1af2c943d8bdfaf19fe
+    // repaid -30077488379451392498995529 dai, when the borrow balance
+    // was 30077488379451392498995503
```
Comment (jaimehgb) [inline subgraphs/makerdao/src/common/helpers.ts:635]: Is this comment still relevant? If we `log.critical` when balance < 0, then we expect it to not happen?

File: subgraphs/makerdao/protocols/makerdao/config/templates/makerdao.template.yaml [hunk 3]
```diff
@@ -50,6 +50,10 @@ dataSources:
         - event: LogNote(indexed bytes4,indexed bytes32,indexed bytes32,indexed bytes32,bytes)
           topic0: "0x7608870300000000000000000000000000000000000000000000000000000000"
           handler: handleVatFrob
+        ### CDP Confiscation
+        - event: LogNote(indexed bytes4,indexed bytes32,indexed bytes32,indexed bytes32,bytes)
```
Comment (jaimehgb) [inline subgraphs/makerdao/protocols/makerdao/config/templates/makerdao.template.yaml:54]: The fix itself is this part I'm assuming?

File: subgraphs/makerdao/src/common/helpers.ts (hunk 1 above)
Comment (tnkrxyz) [reply subgraphs/makerdao/src/common/helpers.ts:514]: Fixed.

File: subgraphs/makerdao/src/common/helpers.ts (hunk 2 above)
Comment (tnkrxyz) [reply subgraphs/makerdao/src/common/helpers.ts:635]: Yes, as a side effect of fixing the index error, this also fixes the small negative position balance issue.

File: subgraphs/makerdao/protocols/makerdao/config/templates/makerdao.template.yaml (hunk 3 above)
Comment (tnkrxyz) [reply subgraphs/makerdao/protocols/makerdao/config/templates/makerdao.template.yaml:54]: Correct.

**Review by jaimehgb — APPROVED** [review]

**Comment by tnkrxyz** [comment]
@jaimehgb , can you re-review this again? My fix didn't work when syncing from scratch. I had to add code to skip handling grab() for liquidations.

- new deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/test-subgraph (sync from scratch)

**Review by jaimehgb — APPROVED** [review]
LGMT

**Comment by tnkrxyz** [comment]
> LGMT

thanks. I will merge once the sync complete without error.

**Pull request merged**